wallet/txauthor/author.go

```diff
@@ -168,6 +170,106 @@ func NewUnsignedTransaction(outputs []*wire.TxOut, feeRatePerKb btcutil.Amount,
 	}
 }
 
+// NewUnsignedTransaction2 is an experimental function to solve the fee-calculation problem which is appareant
+// in NewUnsignedTransaction.
+func NewUnsignedTransaction2(outputs []*wire.TxOut, feeRatePerKb btcutil.Amount,
+	inputs []wtxmgr.Credit, selectionStrategy InputSelectionStrategy, changeSource *ChangeSource) (*AuthoredTx, error) {
+
+	changeScript, err := changeSource.NewScript()
+	if err != nil {
+		return nil, err
+	}
+
+	targetAmount := SumOutputValues(outputs)
+
+	inputSelection := txSelector{
+		inputState: &inputState{
+			feeRatePerKb:      feeRatePerKb,
+			targetAmount:      targetAmount,
+			outputs:           outputs,
+			selectionStrategy: selectionStrategy,
+			changeOutpoint: wire.TxOut{
+				// We start from the negative amount to make
+				// sure the initial add of the first input
+				// succeeds and does not fail with negative
+				// yielding.
+				Value:    -int64(targetAmount),
```

Comment:
Thanks for your thought, could you elaborate more, not sure if we can compare to the currentAmount, because we have to make sure we look at the fees differently, but happy if you can tell me more details how you envisioned it.

I found another solution currently, keeping the changeOutput at 0 but checking in the totalOutput calculation whether we have already some funds in your input array if not returen 0.

```
// totalOutput is the total amount left for us after paying fees.
//
// NOTE: This might be dust.
func (t *inputState) totalOutput() btcutil.Amount {
	// We return an output amount of 0 so that the first
	// input is added successfully. Otherwise adding the
	// first amount would fails as long its smaller than
	// the target amount.
	if len(t.inputs) == 0 {
		return 0
	}

	return t.targetAmount + btcutil.Amount(t.changeOutpoint.Value)
}
```

Reply:
What I meant is that:

in `addToState` instead of:

```
tempInputState.changeOutpoint.Value = int64(tempInputState.inputTotal - tempInputState.targetAmount - fee)
```
we can have something like

`currentAmount = tempInputState.inputTotal - fee(with change)` + Check if yield > 0 

You can create a new function `getChangeOutput` (that will be currentAmount-targetAmount) to get the theoretical changeOutput every time you need (to check if dust or not,...).

We can argue however that the changeOutput is still negative 😄  But for the yield, we can only check
`currentAmount - currentAmount_old` (?)
Finally, that's what you did with the `totalOutput` because `targetAmount` is always the same. 
So, when you check the yield, you simply check `(inputTotal-fee) - (inputTotal-fee)_old` AKA `currentAmount - currentAmount_old`.

Maybe the name `currentAmount` is misleading and could be something more explicit.

However, I do agree that it could be a good idea to have a behaviour similar to lnd (so using a changeOutput field like you did) but without it being explicitly negative from start. (This logic has proven itself in lnd) 😄 

That's why maybe your solution can be a good idea (I haven't check in details your new solution though)

Maybe it's only me that doesn't like the change output being explicitly negative from start! (only to solve a problem when adding the first input)

Reply:
(ah, you just changed your solution, will check again)

Reply:
Thanks for laying out your reasoning. I think you solution might also be good, to have a kind of a theoretical change. At the end all boils down to the initial addition of the first output. 

I think just capping the totalAmount at 0 when there is still no input in our state might be the easier solution regarding the amount of source code haha. But your solution might be more verbose lets see what the other reviewers say and then go from there ?